features/filesystem/bd/ChainingBlockDevice.cpp

```diff
@@ -0,0 +1,210 @@
+            _erase_size = _erase_size;
+        } else {
+            return BD_ERROR_PARAMETER;
+        }
+
+        _size += _bds[i]->size();
+    }
+
+    return 0;
+}
+
+bd_error_t ChainingBlockDevice::deinit()
+{
+    for (size_t i = 0; i < _bd_count; i++) {
+        bd_error_t err = _bds[i]->deinit();
+        if (err) {
+            return err;
+        }
+    }
+
+    return 0;
+}
+
+bd_error_t ChainingBlockDevice::read(void *b, bd_addr_t addr, bd_size_t size)
+{
+    if (!is_valid_read(addr, size)) {
+        return BD_ERROR_PARAMETER;
+    }
+
+    uint8_t *buffer = static_cast<uint8_t*>(b);
+
+    // Find block devices containing blocks, may span multiple block devices
+    for (size_t i = 0; i < _bd_count && size > 0; i++) {
+        bd_size_t bdsize = _bds[i]->size();
+
+        if (addr < bdsize) {
+            bd_size_t read = size;
+            if (addr + read > bdsize) {
+                read = bdsize - addr;
+            }
+
+            bd_error_t err = _bds[i]->read(buffer, addr, read);
+            if (err) {
+                return err;
+            }
+
+            buffer += read;
+            addr += read;
+            size -= read;
+        }
+
+        addr -= size;
+    }
+
+    return 0;
+}
+
+bd_error_t ChainingBlockDevice::program(const void *b, bd_addr_t addr, bd_size_t size)
+{
+    if (!is_valid_program(addr, size)) {
+        return BD_ERROR_PARAMETER;
+    }
+
+    const uint8_t *buffer = static_cast<const uint8_t*>(b);
+
+    // Find block devices containing blocks, may span multiple block devices
+    for (size_t i = 0; i < _bd_count && size > 0; i++) {
+        bd_size_t bdsize = _bds[i]->size();
+
+        if (addr < bdsize) {
+            bd_size_t program = size;
+            if (addr + program > bdsize) {
+                program = bdsize - addr;
+            }
+
+            bd_error_t err = _bds[i]->program(buffer, addr, program);
+            if (err) {
+                return err;
+            }
+
+            buffer += program;
+            addr += program;
+            size -= program;
+        }
+
+        addr -= size;
+    }
+
+    return 0;
+}
+
+bd_error_t ChainingBlockDevice::erase(bd_addr_t addr, bd_size_t size)
+{
+    if (!is_valid_erase(addr, size)) {
+        return BD_ERROR_PARAMETER;
+    }
+
+    // Find block devices containing blocks, may span multiple block devices
+    for (size_t i = 0; i < _bd_count && size > 0; i++) {
+        bd_size_t bdsize = _bds[i]->size();
+
+        if (addr < bdsize) {
+            bd_size_t erase = size;
+            if (addr + erase > bdsize) {
+                erase = bdsize - addr;
+            }
+
+            bd_error_t err = _bds[i]->erase(addr, erase);
+            if (err) {
+                return err;
+            }
+
+            addr += erase;
+            size -= erase;
+        }
+
+        addr -= size;
+    }
+
+    return 0;
+}
+
+bd_size_t ChainingBlockDevice::read_size()
```

Comment:
These are the only existing function I could find:
- [CallChain::size](https://github.com/ARMmbed/mbed-os/blob/master/platform/CallChain.h#L137)
- [FileHandle::flen](https://github.com/ARMmbed/mbed-os/blob/master/drivers/FileHandle.h#L107)

The `size` function is the odd one because of the STL. Maybe keep `size` but use `get_` for `read/write/erase_size`? Thoughts?